### ISSUE TITLE
feat: search and install Schematics

### ISSUE THREAD

**Issue by manekinekko** [angular/angular-cli#10000]
## RFC 

The Angular CLI has introduced support for [Schematics](https://github.com/angular/devkit) allowing developers to easily extend or even completely change the default behavior of the Angular CLI commands. With Schematics, we can (for instance)...

- use the official [`ng update`](https://www.npmjs.com/package/@schematics/package-update) command to update the Angular dependencies of a project (and more).
- use the [Nrwl Nx Workspace](https://nrwl.io/nx) to scaffold enterprise-grade Angular applications with proven project structure and patterns.
- use your own Schematics to transform any [Feature module](https://angular.io/guide/feature-modules) to a [Lazy Loading Feature module](https://angular.io/guide/lazy-loading-ngmodules).

More generally, you can use Schematics to "[...] transforms to your project, such as create a new component or updating your code to fix breaking changes in a dependency. Or maybe you want to add a new configuration option or framework to an existing project". Read more on the [angular.io](https://blog.angular.io/schematics-an-introduction-dc1dfbc2a2b2).

The Angular community has already started creating custom Schematics, here are some of them:

- [@nrwl/schematics](https://github.com/nrwl/nx)
- [@ngrx/schematics](https://github.com/ngrx/platform/blob/master/docs/schematics/README.md)
- [@ionic-angular/schematics](https://github.com/Robinyo/ionic-angular-schematics)
- [@nativescript/schematics](https://github.com/nativescript/nativescript-schematics#readme)
- [@angular-extensions/schematics](https://github.com/angular-extensions/schematics#readme)
- [@ssilvert/keycloak-schematic](https://github.com/ssilvert/keycloak-schematic#readme)
- [@decerto/schematics](https://github.com/decerto-poland/decerto-schematics#readme)
- [@michael.warneke/ng-schematics](https://github.com/MichaelWarneke/ng-schematics#readme)
- [material-schematics](https://github.com/amcdnl/material-schematics)
- [ngx-schematics-collection](https://github.com/wellwind/ngx-schematics-collection)
- [angular-crud](https://github.com/manfredsteyer/angular-crud#readme)
- [custom-schematics](https://github.com/gioboa/custom-schematics#readme)
- [ng-schematics-utils](https://github.com/caroso1222/ng-schematics-utils#readme)
- [enterprise-angular-generator](https://github.com/cambronjay/enterprise-angular-generator#readme)

We expect the number of the Angular CLI Schematics to grow; hence, the goal of this thread is to discuss the ability to add a new `ng` command allowing to (right from the command line):

#### Note: the commands used below `ng schematics **` are just for illustration purposes!

0. Searching for available Schematics

```
> ng schematics search foo_
● foo_abc     (description)      (popularity index)
○ foo_def     (description)      (popularity index)   
○ foo_ghi     (description)      (popularity index)
○ foo_ijk     (description)      (popularity index)
```

1. Interactively search for Schematics

```
> ng schematics search
Enter keyword: foo_
● foo_abc     (description)      (popularity index)
○ foo_def     (description)      (popularity index)   
○ foo_ghi     (description)      (popularity index)
○ foo_ijk     (description)      (popularity index)
```

2. Interactively installing Schematics

```
> ng schematics search 
Enter keyword: bar_
● bar_def     (description)      (popularity index)   
○ bar_ijk     (description)      (popularity index)

Selected:
● foo_abc     1.2.3
● foo_ghi     0.1.2
```

3. Install Schematics

```
> ng schematics install foo_abc foo_ghi
Installing...
```

In order for this search to be done, we need to be able to identify what NPM packages are actually Angular CLI Schematics. Here are different proposals:

1. Use an NPM named scope, eg. `@ngxschematics` (`ngx` being the community reserved prefix we agreed on (cc @IgorMinar) — `@schematics` is already reserved for the official Schematics made by the Angular team). The community NPM scope (`@ngxschematics`) would obviously be owned (and maintained) by the Angular core team. So (community) Schematics authors will have to send PRs to that repository to make their Schematics available (listed) in the Angular CLI search command (discussed here). This approach is similar to the [DefinitelyTyped](http://definitelytyped.org/) project (`@types/**`).

2. Have some sort of a "Market Place for Schematics" (out of the scope of this proposal). This could be merged with the previous solution.

3. Community Schematics authors would have to tag their schematics with an explicit set of tags, eg. `(angular, schematics)`. Community Schematics that aren't tagged using these keywords won't be listed by the Angular CLI search command (discussed here).

The search process would be done by using 3rd party search APIs, here are some of them:

1. [npms.io](https://npms.io/about)
2. [npmsearch.com](http://npmsearch.com/)
3. (github.com/algolia/npm-search)(https://github.com/algolia/npm-search)

<hr/>

Similar projects (inspiration):

- [https://github.com/maticzav/emma-cli](https://github.com/maticzav/emma-cli)
- [https://microsoft.github.io/TypeSearch/](https://microsoft.github.io/TypeSearch/)
- [Vue CLI search command](https://github.com/vuejs/vue-cli/commit/83939c9214609fe2db4d24a7f1f4b90f369355d2)

<hr/>

Please use Github's emoji to express your interest in this draft.

cc @hansl @Brocco @filipesilva

**Comment by manfredsteyer** [comment]
I like those ideas. Two things came in mind when reading this:
- As Schematic Collections are distributed via npm packages, we could leverage the existing npm infrastructure
- There is not always a strict separation b/w Schematics and libraries as a library can come with its own Schematics baked in

**Comment by vitaliy-bobrov** [comment]
Great initiative, regarding search IMHO tags, looks better, as the simplier way to be listed in the search results. More steps to follow with publication === more contributions.

**Comment by manekinekko** [comment]
@manfredsteyer 

> As Schematic Collections are distributed via npm packages, we could leverage the existing npm infrastructure

Agreed. Although, this will depend on what DX we wanna have (simple query search, autocomplete...etc). 

> There is not always a strict separation b/w Schematics and libraries as a library can come with its own Schematics baked in

This proposal is only for Schematics that are published on NPM. It's not meant for Schematics

**Comment by manekinekko** [comment]
@vitaliy-bobrov I okay with that too ^^

**Comment by choucry13** [comment]
@manekinekko , it's very a good idea and if we use a specific schematics in enterprise, we can store it on a private nexus (a hoster for all packager type like npm, nuget, maven...). Do you think it's possible or not ?

**Comment by Haroenv** [comment]
Hey! Just chiming in here from Algolia, you can see a similar search made with our index in [Emma-cli](https://github.com/maticzav/emma-cli). It's made for the [Yarn website](https://yarnpkg.com) originally, but definitely also meant to be used by other parts of the community. 

We'd prefer it if the search said `powered by Algolia` or something near the results (can be underneath or to the right edge maybe?), because we provide it as a free service and want as many people/projects as possible to know about it.

![Emma cli demo](https://github.com/maticzav/emma-cli/raw/master/media/example.gif)

Ping me if you have any questions or want to try it out

**Comment by Toxicable** [comment]
I made another little here https://schematics-search.firebaseapp.com/
It's search is pretty crude but it somewhat works and there's lots of optimizations that could be done to speed it up.
The main point here is showing the kinds of speed it would take at max to find packages that are usable as Schematics

**Comment by StephenFluin** [comment]
Love this. I've been recommending people to add specific tags including:

`ng-add` `ng-update` `ng-generate` `ng-new` to indicate their capabilities.
